### ISSUE TITLE
Entities MUST contain a rel attribute?

### ISSUE THREAD

**Issue by ericelliott** [kevinswiber/siren#17]
> entities

> A collection of related sub-entities. If a sub-entity contains an href value, it should be treated as an embedded link. Clients may choose to optimistically load embedded links. If no href value exists, the sub-entity is an embedded entity representation that contains all the characteristics of a typical entity. **One difference is that a sub-entity MUST contain a rel attribute to describe its relationship to the parent entity.**

I'm building a resourceful API. Each resource has an index where you can list / search / browse child entities.

The way that it's structured, it's obvious that the relationship of the entities in the index is that they are child records associated with the resource. For example:

```
/albums
Nevermind - Nirvana
Pretty Hate Machine - Nine Inch Nails
Discovery - Daft Punk
```

I don't want to come up with a different rel relationship for each kind of resource, or it would turn into an ever-proliferating taxonomy that essentially adds no real value.

Which leads me to my next point: These rels aren't adding any value to my API, but there's a rel associated with every record in the list -- the *same* rel. It doesn't make any sense to duplicate that rel for every item in the list.

I'd just make these links instead (SHOULD, yay!), except then I can't specify properties to display with each link (a nice feature of entities which lets you call out the interesting features of an entity which might be good for listing entities on a web page).

Any chance we could change MUST to SHOULD here? I'm not sure I understand the rigidity of this requirement.

**Comment by ericelliott** [comment]
The more I think about this, the more I'm convinced that MUST is too strong. Requiring sub-entities to contain a `rel` to its parent should really be the parent collection's concern. Sub entities can and most likely should be blissfully unaware that a parent collection even exists.

Requiring them to embed data relating to the parent collection is redundant, wasteful, and could cause tight coupling between children and parent by implication.

`rel` makes sense for links in the context of a parent document describing the relationship of the linked entity to the parent entity. It does not make sense in the context of a child which may have many parent entities. That mimics the tight coupling in classical inheritance which [I have ranted about in detail](https://medium.com/javascript-scene/the-two-pillars-of-javascript-ee6f3281e7f3), elsewhere.

I feel quite strongly about this. If it is not a change we're going to make in Siren, I intend to fork Siren and implement a competing standard. I'm already ignoring this requirement in all my implementations, optimistically hoping we can remove this requirement.

I do support `entityAttributes` in all my implementations which would allow clients to figure out any common relationships in all the entities.

**Comment by ericelliott** [comment]
https://github.com/kevinswiber/siren/pull/45

**Comment by apsoto** [comment]
I think an entity should be an addressable resource, but the error case is a bit of a special case, so to use that as the argument for no `rels` on the sub-entity doesn't seem valid IMO.  Also, throwing the fork word around if you don't get your way doesn't help with the discussion.  Maybe it makes more sense to work together towards a Siren 2.0 that is not backwards compatible.

However, having the `rel` on the sub-entity has caused me a little extra work because the sub-entity response JSON is different from the same resource's top-level entity's response JSON.

For example:

    /series/:id/episodes/:id
    /episodes/:id

In this case, it's the same resource, just the specific **episode** is also a sub-entity to the **series**.  I'd love to be able to re-use the response JSON for the `/episodes/:id` when building the **series** response (and vice versa).  However, having the `rel` attribute on the sub-entity means I can't.

**Comment by ericelliott** [comment]
> Also, throwing the fork word around if you don't get your way doesn't help with the discussion.

I need this change, so if it doesn't happen here, it will happen in a fork. I'm not making threats, just making my intentions clear.

So far I haven't heard any convincing arguments that we MUST include rel attributes in sub-entities. They're already required for sub-entity links, and that makes much more sense.

Why not make the sub-entity requirements the same as a root level entity? =)

> having the rel on the sub-entity has caused me a little extra work because the sub-entity response JSON is different from the same resource's top-level entity's response JSON.

Another compelling reason to remove "MUST".

=)

**Comment by kevinswiber** [comment]
This issue has been hanging around for a while, and I think I'm still having trouble seeing why you would need a sub-entity without a relation value.  I am open to seeing it this way, but chances are, I just need a little push.  :)

Part of this might be because I often design my inline sub-entities as a subset of their full representation.  (Think of a list view representation vs. a complete item representation.)  It's a slightly different model, and therefore, I need to have specific code to piece together the representation anyway, often in the parent context.  Adding a link relation is no big deal in this case.

Here's my big question: As a client, how do I determine why I should care about the sub-entity?  I currently use the link relation for that.  If there were no link relation, what information should I inspect?

I think we might be able to get better progress on this with the magic power of voice, video, and screen-sharing. @ericelliott Are you open to discussing this on the Siren Hangout this Friday? We can start piecing together an agenda.

**Comment by kevinswiber** [comment]
Also, regarding forking...  I'm completely open and supportive of anyone looking to fork any work I've touched.  I think of that as a positive action.  That said, I'm hoping we can foster an open community who feels that Siren is, in fact, a collectively owned project.  If we all arrive at an understanding of the change request and there's a legitimate disagreement about how Siren should move forward, I think that would be a good case for forking, and I would offer my full support.  Otherwise, I would love to see that energy put into making the Siren specification better for all of us.  :)

**Comment by ericelliott** [comment]
I'm all for improving Siren instead of forking. I'd love to discuss it with you in a hangout. Have we settled on a specific time and date?

I'm also completely open to the idea of having relations for all sub-entities. My problem with it is where those relationships are located. Almost all of my use cases for entities boil down to collections of similar things, like lists of albums, lists of orders, lists of users, lists of `x`, and in all of those cases, the collection -- not the items themselves -- are responsible for keeping track of the relationship, which all items share.

So my problem is 2-fold:

* Including rel at the sub-item level is redundant and wasteful, because all items share the same rel to the parent.
* We can't just reuse the same representation that the item has at the root level of a response, meaning that we'd have to inject the rel to produce the sub-entity collections, causing us to write special case code, when we could have just used exactly the same code to produce the sub-entities.

In other words, this requirement forces more software complexity on the API implementation, and creates larger payloads, and it offers absolutely no value that couldn't be achieved with a simpler implementation, such as a shared `entityAttributes` key in the parent entity.

**Comment by uglymunky** [comment]
> I think an entity should be an addressable resource

+1 I feel this distinction is key to differentiating between what is an entity and what is a property.

> As a client, how do I determine why I should care about the sub-entity? I currently use the link relation for that. If there were no link relation, what information should I inspect?

What about using the properties object? (I didn't come up with this but remember seeing this mentioned on a thread a while back)

ex:
```
	{
		properties: {
			name: 'ralph'
			, age: 78
			, profile_image: 'self://x.io/user/1234/image'
			, friends: 'self://x.io/user/1234/friends'
		}
		
		, entities: [
			{ /* "profile_image" entity */ }
			, { /* "friends" collection entity */ }
		]
	}
```

We could discuss the proper syntax for the internal reference (e.g. the `self://`), as well as how to treat linked sub-entities, and any other details in a separate thread but I think you get the idea.

* Retains the relationship information from parent to child
* Allows you to re-use the root representation without having to modify it
* Collections don't need a rel so this would not affect collections.
* Allows for clients to implement hash fragments (e.g. http://x.io/user/1234#profile_image)
* Would resolve this old issue: https://github.com/kevinswiber/siren/pull/33 :)

Also, wrt @apsoto's comments in #45:

> The entities property is currently an array, so it's a breaking change to convert it to an object

This is backward compatible

> In addition, current spec requires the rels to be an array. Using the example above means there is only one rel per child entity.  I don't think I'd miss that requirement very much, but I only have the perspective of my specific API's problem domain.

I also would not miss that requirement :)

> I like treating everything as a collection instead of having the possibility of it being an object or an array

This keeps the entities array an array.

**Comment by ericelliott** [comment]
Seems somebody beat me to changing this requirement. I don't see it in the current version.

Should we close this, or do we still need to add wording for the default `rel` we discussed in the hangout?

**Comment by ericelliott** [comment]
Oops. Looks like I jumped the gun.. now I'm looking at the doc again, I still see the requirement. I'll open a PR.
